src/lib/Inference.hs

```diff
@@ -241,6 +241,10 @@ checkOrInferRho (WithSrc pos expr) reqTy =
   UHole -> case reqTy of
     Infer -> throw MiscErr "Can't infer type of hole"
     Check ty -> freshType ty
+  UAnnotatedExpr val ty -> do
+    ty' <- checkUType ty
+    val' <- checkRho val ty'
```

Comment:
I think that this could actually be `checkOrInferRho val Concrete ty'`, which would let us e.g. do better inference for table literals.

Reply:
That's something I don't quite understand. What is the difference between`Concrete ty` and `Suggest ty`?

Reply:
In particular I notice that regular binders sometimes use `Concrete` and sometimes use `Suggest`? https://github.com/google-research/dex-lang/blob/208584d880d6406f35ad1da76edd8e7d9d8003cc/src/lib/Inference.hs#L359

Also, I wonder if this should call `checkSigma` instead of `checkOrInferRho`. Maybe an explicit annotation should act the same way as a `UDecl`.

Reply:
I've made it match `UDecl`; let me know if that makes sense.

Reply:
Yes, that's perfect. The difference between `Check` and `Concrete` is that in `Concrete` checking mode you have a guarantee that the type we're pushing down has no inference variables inside it, while `Check` is a best effort thing. For example, in `Check`ing mode you can find only a partially unified type (e.g. `(n:Type) ?-> (n, ?1)`) or even a completely non-informative inference variable you know nothing about.

I've been recently working on type inference for dependent table literals and I realized that to do it correctly, we'd either need to implement some form of higher order unification or the `Concrete` workaround. In the first case, we'd unify the table type with `i:?1 -> ?2 i` (note that `?2` is a function!) and proceed. Otherwise, we have no way of actually allowing dependency on `i`, unless it is clearly visible. Anyway, that is complicated, so it is just a lot simpler to allow typing dependent tables only when they're explicitly annotated with a type, and in that case we know all the details about dependency.